modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala

```diff
@@ -111,31 +115,45 @@ object ParsedConfigs {
           }
       case OutputConfig.Kinesis(_, _, Some(key), _, _, _) if !enrichedFieldsMap.contains(key) =>
         NonEmptyList.one(s"Partition key $key not valid").invalid
+      case ka: OutputConfig.Kafka if !ka.headers.forall(enrichedFieldsMap.contains) =>
+        NonEmptyList
+          .one(
+            s"Fields [${ka.headers.filterNot(enrichedFieldsMap.contains).mkString(", ")}] for headers are not part of the enriched event"
+          )
+          .invalid
       case _ =>
         output.valid
     }
 
   private[config] def outputAttributes(output: OutputConfig): EnrichedEvent => Map[String, String] =
     output match {
-      case OutputConfig.PubSub(_, Some(attributes), _, _, _) =>
-        val fields = ParsedConfigs.enrichedFieldsMap.filter {
-          case (s, _) =>
-            attributes.contains(s)
-        }
-        attributesFromFields(fields)
-      case OutputConfig.Kinesis(_, _, Some(key), _, _, _) =>
-        val fields = ParsedConfigs.enrichedFieldsMap.filter {
-          case (s, _) =>
-            s == key
-        }
-        attributesFromFields(fields)
-      case _ =>
-        _ => Map.empty
+      case OutputConfig.PubSub(_, Some(attributes), _, _, _) => attributesFromFields(attributes)
+      case OutputConfig.Kinesis(_, _, Some(key), _, _, _) => attributesFromFields(Set(key))
+      case OutputConfig.Kafka(_, _, _, headers, _) => attributesFromFields(headers)
+      case _ => _ => Map.empty
+    }
+
+  private[config] def attributesFromFields(attributes: Set[String]): EnrichedEvent => Map[String, String] = {
+    val fields = ParsedConfigs.enrichedFieldsMap.filter {
+      case (s, _) =>
+        attributes.contains(s)
+    }
+    (ee: EnrichedEvent) =>
+      fields.flatMap {
+        case (k, f) =>
+          Option(f.get(ee)).map(v => k -> v.toString)
+      }
+  }
+
+  private[config] def outputPartitionKey(output: OutputConfig): EnrichedEvent => String =
+    output match {
+      case OutputConfig.Kafka(_, _, partitionKey, _, _) => partitionKeyFromFields(partitionKey)
+      case OutputConfig.Kinesis(_, _, Some(partitionKey), _, _, _) => partitionKeyFromFields(partitionKey)
+      case _ => _ => ""
     }
 
-  private def attributesFromFields(fields: Map[String, Field])(ee: EnrichedEvent): Map[String, String] =
-    fields.flatMap {
-      case (k, f) =>
-        Option(f.get(ee)).map(v => k -> v.toString)
+  private[config] def partitionKeyFromFields(partitionKey: String): EnrichedEvent => String =
+    ParsedConfigs.enrichedFieldsMap.get(partitionKey).fold[EnrichedEvent => String](_ => "") { f => ee =>
```

Comment:
What would it mean if `ParsedConfigs.enrichedFieldsMap` is _missing_ the key `partitionKey`.  Would this mean the user has provided an invalid hocon file? Do you have any validation anywhere to prevent this happening?  I would think if the map is missing the partition key, then it would be better to raise an exception instead of returning `_ => ""`

Reply:
> Would this mean the user has provided an invalid hocon file?

yes and I don't have any validation to prevent it atm. I'll improve output decoder to make sure enrich-kafka starts with a valid partitionKey and a valid set of headers so that we don't need to raise an exception